Core/Barcelona/Extensions/Glue/Publishers.swift

```diff
@@ -0,0 +1,76 @@
+//
+//  Publishers.swift
+//  Barcelona
+//
+//  Created by June Welker on 2/15/23.
+//
+
+import Foundation
+import Combine
+
+extension Publisher {
+    func toFuture() -> Future<Output, Failure> {
+        return Future<Output, Failure> { resolve in
+            self.retainingSink {
+                if case let .failure(error) = $0 {
+                    resolve(.failure(error))
+                }
+            } receiveValue: {
+                resolve(.success($0))
+            }
+        }
+    }
+
+    @discardableResult
+    func retainingSink(receiveCompletion: @escaping (Subscribers.Completion<Failure>) -> Void, receiveValue: @escaping (Output) -> Void) -> AnyCancellable? {
+        var cancellable: AnyCancellable?
+        cancellable = sink(receiveCompletion: {
+            receiveCompletion($0)
+
+            withExtendedLifetime(cancellable) { cancellable = nil }
+        }, receiveValue: receiveValue)
+
+        return cancellable
+    }
+
+    @discardableResult
+    func retainingAwaitSingleOutput(_ onComplete: @escaping (Result<Output, Failure>) -> Void) -> AnyCancellable? {
+        var cancellable: AnyCancellable?
+        cancellable = retainingSink {
+            if case let .failure(error) = $0 {
+                onComplete(.failure(error))
+            }
+        } receiveValue: {
+            onComplete(.success($0))
+            cancellable?.cancel()
+        }
+
+        return cancellable
+    }
+}
+
+extension Publisher where Failure == Never {
+    @discardableResult
+    func retainingAwaitSingleOutput(_ onComplete: @escaping (Output) -> Void) -> AnyCancellable? {
+        retainingAwaitSingleOutput {
+            // We know that this is fine since the failure can't be constructed
+            if case let .success(output) = $0 {
+                onComplete(output)
+            }
+        }
+    }
+}
+
+extension Sequence {
+    func asyncMap<T>(_ transform: @escaping (Element) async -> T) async -> [T] {
+        await withTaskGroup(of: T.self, returning: [T].self) { group in
+            for item in self {
+                group.addTask {
+                    await transform(item)
+                }
+            }
+
+            return await group.reduce(into: []) { $0.append($1) }
+        }
+    }
+}
```

Comment:
Are these part of this PR? I can't see them used anywhere.

Additionally, I'm not really a fan of the magic retaining sink. Creating a `var cancellables = Set<AnyCancellable>()` ensures predictable lifetimes, auto cancellation on `deinit` etc., we should really favor those.

Reply:
Oh yeah, I was using these during development but I'm not certain all of them actually ended up in the PR. I'll fix that.

And I used to be a fan of using a set, but I feel like it's not applicable for all circumstances. For example, a lot of functions in barcelona don't have an associated class (they're just global functions) and thus can't utilize a set of cancellables. Plus, if you use a set for cancellables that aren't supposed to be long-running (e.g. a single read from the database, even though we're not using combine for that), you can fill up the set with cancellables that aren't wanted anymore and cause a small memory leak. I think the set pattern should only be used in the case of pipelines/cancellables that should stay alive during the entire lifetime of the object.

Reply:
Good points 👍 I'd still favor anything explicit, though, maybe with properties if there's a concern for filling up the set (which is valid in some cases and one I never understood about Combine's design). For global functions, just forwarding the pipeline enough for something to save it properly would be the best, but really I don't feel like demanding anything "best" with this codebase right now.

But let's see how to approach those the next time something comes up. This is probably a meaningless discussion anyway, as I feel concurrency is a much better fit here for many reasons.